opencodeblocks/blocks/codeblock.py

```diff
@@ -120,11 +159,247 @@ def run_code(self):
+                        input_block = edge.source_socket.block
+                        to_visit_input.append(input_block)
+                        if input_block not in visited:
+                            next_blocks.append(input_block)
+                        visited.append(input_block)
+                to_visit_input.remove(block)
+            for block in to_visit_output.copy():
+                # Check output edges and blocks
+                for output_socket in block.sockets_out:
+                    for edge in output_socket.edges:
+                        if edge not in visited:
+                            next_edges.append(edge)
+                        visited.append(edge)
+                        output_block = edge.destination_socket.block
+                        to_visit_input.append(output_block)
+                        to_visit_output.append(output_block)
+                        if output_block not in visited:
+                            next_blocks.append(output_block)
+                        visited.append(output_block)
+                to_visit_output.remove(block)
+
+            # Add the next stage to to_transmit
+            to_transmit.append(next_edges)
+            to_transmit.append(next_blocks)
+
+            # Reset next stage
+            next_edges = []
+            next_blocks = []
+
+        return to_transmit
+
+    def run_blocks(self):
+        """Run a list of blocks"""
+        for block in self.blocks_to_run[::-1]:
+            if not block.has_been_run:
+                block.run_code()
+        if not self.has_been_run:
+            self.run_code()
+
+    def run_left(self):
+        """Run all of the block's dependencies and then run the block"""
+
+        # Reset has_been_run to make sure that the self is run again
+        self.has_been_run = False
+
+        # To avoid crashing when spamming the button
+        if len(self.transmitting_queue) != 0:
+            return
+
+        # Gather dependencies
+        blocks_to_run, to_transmit = self.custom_bfs(self)
+        self.blocks_to_run = blocks_to_run
+
+        # Set the transmitting queue
+        self.transmitting_queue = to_transmit
+        # Set delay so that the transmitting animation has fixed total duration
+        self.transmitting_delay = int(
+            self.transmitting_duration / len(self.transmitting_queue)
+        )
+        # Start transmitting animation
+        self.transmitting_animation_in()
+
+    def run_right(self):
+        """Run all of the output blocks and all their dependencies"""
+
+        # To avoid crashing when spamming the button
+        if len(self.transmitting_queue) != 0:
+            return
+
+        # Create transmitting queue
+        self.transmitting_queue = self.custom_djikstra()
+
+        # Gather outputs
+        blocks_to_run, _ = self.custom_bfs(self, reverse=True)
+        self.blocks_to_run = blocks_to_run
+
+        # For each output found
+        for block in blocks_to_run[::-1]:
```

Comment:
Why are multiple BFS needed. This loop seems to implement a BFS itself ?
Shouldn't the BFS return all the blocks to run in one pass ?

Reply:
BFS will do traversal in only one direction, either towards the inputs or towards the outputs. We need an output-BFS to gather the outputs of self and we need to do input-BFS on all of the output blocks to gather dependencies for each